consensus/core/src/synchronizer.rs

```diff
@@ -0,0 +1,480 @@
+    /// the peer authority.
+    pub async fn fetch_blocks(
+        &self,
+        block_refs: BTreeSet<BlockRef>,
+        peer_index: AuthorityIndex,
+    ) -> Result<(), ConsensusError> {
+        let (sender, receiver) = oneshot::channel();
+        self.commands_sender
+            .send(Command::FetchBlocks {
+                missing_block_refs: block_refs,
+                peer_index,
+                result: sender,
+            })
+            .await
+            .map_err(|_err| ConsensusError::Shutdown)?;
+        receiver.await.map_err(|_err| ConsensusError::Shutdown)?
+    }
+
+    pub async fn stop(&self) {
+        let mut tasks = self.tasks.lock();
+        tasks.abort_all();
+    }
+}
+
+#[allow(dead_code)]
+pub(crate) struct Synchronizer {
+    context: Arc<Context>,
+    commands_receiver: Receiver<Command>,
+    fetch_block_senders: BTreeMap<AuthorityIndex, Sender<BTreeSet<BlockRef>>>,
+}
+
+impl Synchronizer {
+    pub fn start<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher>(
+        network_client: Arc<C>,
+        context: Arc<Context>,
+        core_dispatcher: Arc<D>,
+        block_verifier: Arc<V>,
+    ) -> Arc<SynchronizerHandle> {
+        let (commands_sender, commands_receiver) = channel(1_000);
+
+        // Spawn the tasks to fetch the blocks from the others
+        let mut fetch_block_senders = BTreeMap::new();
+        let mut tasks = JoinSet::new();
+        for (index, _) in context.committee.authorities() {
+            if index == context.own_index {
+                continue;
+            }
+            let (sender, receiver) = channel(FETCH_BLOCKS_CONCURRENCY);
+            tasks.spawn(Self::fetch_blocks_from_authority(
+                index,
+                network_client.clone(),
+                block_verifier.clone(),
+                context.clone(),
+                core_dispatcher.clone(),
+                receiver,
+            ));
+            fetch_block_senders.insert(index, sender);
+        }
+
+        // Spawn the task to listen to the requests & periodic runs
+        tasks.spawn(async {
+            let mut s = Self {
+                context,
+                commands_receiver,
+                fetch_block_senders,
+            };
+            s.run().await;
+        });
+
+        Arc::new(SynchronizerHandle {
+            commands_sender,
+            tasks: Mutex::new(tasks),
+        })
+    }
+
+    async fn fetch_blocks_from_authority<
+        C: NetworkClient,
+        V: BlockVerifier,
+        D: CoreThreadDispatcher,
+    >(
+        peer_index: AuthorityIndex,
+        network_client: Arc<C>,
+        block_verifier: Arc<V>,
+        context: Arc<Context>,
+        core_dispatcher: Arc<D>,
+        mut receiver: Receiver<BTreeSet<BlockRef>>,
+    ) {
+        const REQUEST_TIMEOUT: Duration = Duration::from_millis(500);
+        const MAX_RETRIES: u32 = 5;
```

Comment:
Kept it simple for now.  Keep in mind this is a best effort approach thus I do attempt a max retry. The periodic synchronization mechanism will guarantee the liveness as backup.